### PR TITLE
Backport lazy collection init

### LINKED ISSUE #3596: [Backport] Lazy collection init
### Original feature issue

- Issue: #3360
- PR: #3363

### Target release

- [x] release33
- [ ] release32
- [ ] release30

### Reason

The SQLite backend support backport request is related to the performance desired that surfaced in https://github.com/cobbler/cobbler/discussions/3578.

### PR DESCRIPTION
**Pull request by SchoolGuy**
## Linked Items

Fixes #3596

## Description

This PR attempts to backport the lazy collection init. This will be off per default.

This is a follow up to #3634 as I sadly can't reopen it.

## Behaviour changes

Old: Cobbler wasn't available until all items are fully loaded.

New: Cobbler will startup up with only the item names in memory until the item is first accessed.

## Category

This is related to a:

- [ ] Bugfix
- [x] Feature
- [ ] Packaging
- [ ] Docs
- [ ] Code Quality
- [ ] Refactoring
- [ ] Miscellaneous

## Tests

- [x] Unit-Tests were created
- [ ] System-Tests were created
- [ ] Code is already covered by Unit-Tests
- [ ] Code is already covered by System-Tests
- [ ] No tests required

### CODE REVIEW

**Comment by SchoolGuy** [comment]
While this PR itself doesn't help I will still let it stay in draft until the collection index is done.

**Comment by SchoolGuy** [comment]
@tpw56j I think this PR is ready for review. As you are the most knowledgeable person about our cache implementation I would love it if you could verify if my backport works as intended.

**Comment by tpw56j** [comment]
@SchoolGuy I'll test the cache implementation on this PR over the weekend.

File: cobbler/cobbler_collections/collection.py [hunk 1]
```diff
@@ -126,9 +133,16 @@ def find(self, name: str = "", return_list: bool = False, no_errors=False,
             except:
                 return self.listing.get(kargs["name"], None)
 
+        if self.api.settings().lazy_start:
+            # Forced deserialization of the entire collection to prevent deadlock in the search loop
+            for obj_name in self.get_names():
+                obj = self.get(obj_name)
+                if obj is not None and not obj.inmemory:
+                    obj.deserialize()
+
```
Comment (tpw56j) [inline cobbler/cobbler_collections/collection.py:142]: This code led to an endless loop when using `mac_address`.
I think we should also backport #3654.

Comment (SchoolGuy) [reply cobbler/cobbler_collections/collection.py:142]: Sounds reasonable. I'll do it as soon as time allows. Did you find anything else?

Comment (tpw56j) [reply cobbler/cobbler_collections/collection.py:142]: This line is not backported:
https://github.com/cobbler/cobbler/blob/da8c2460f388603cc46e8002ece4d0fd0151c9c2/cobbler/items/item.py#L271

**Comment by SchoolGuy** [comment]
@tpw56j I would feel much better if I had the honour that you check if my backport is now complete.

File: cobbler/items/item.py [hunk 2]
```diff
@@ -1252,3 +1286,21 @@ def clean_cache(self, name: Optional[str] = None):
         :param name: The name of Item attribute or None.
         """
         self._clean_dict_cache(name)
```
Comment (tpw56j) [inline cobbler/items/item.py:1289]: In the main branch here is the code:
https://github.com/cobbler/cobbler/blob/da8c2460f388603cc46e8002ece4d0fd0151c9c2/cobbler/items/item.py#L1336-L1337

but I suggest changing to:
```python
if self._inmemory and self.api.settings().cache_enabled:
    self._clean_dict_cache(name)
```
and remove the lines:
https://github.com/cobbler/cobbler/blob/da8c2460f388603cc46e8002ece4d0fd0151c9c2/cobbler/items/item.py#L1311-L1313

**Comment by tpw56j** [comment]
Missing line:
https://github.com/cobbler/cobbler/blob/da8c2460f388603cc46e8002ece4d0fd0151c9c2/cobbler/items/item.py#L871

File: cobbler/items/image.py [hunk 3]
```diff
@@ -127,7 +131,7 @@ def arch(self, arch: Union[str, enums.Archs]):
         """
         self._arch = enums.Archs.to_enum(arch)
 
-    @property
+    @LazyProperty
```
Comment (tpw56j) [inline cobbler/items/image.py:134]: Inheritance is broken here.
```
@InheritableProperty
```

File: cobbler/items/image.py [hunk 4]
```diff
@@ -339,7 +343,7 @@ def network_count(self, network_count: int):
             raise TypeError("Field network_count of object image needs to be of type int.")
         self._network_count = network_count
 
-    @property
+    @LazyProperty
```
Comment (tpw56j) [inline cobbler/items/image.py:346]:
```
@InheritableProperty
```

**Comment by SchoolGuy** [comment]
@tpw56j Better now?

**Comment by tpw56j** [comment]
@SchoolGuy Inheritance is broken in many places in `image.py`, `profile.py`, `system.py`, but perhaps this is a topic for a separate backport PR?

I remembered that to fix inheritance, in addition to changing decorators, we need to change the code in some places, as was done in #3558.

**Comment by SchoolGuy** [comment]
@tpw56j I think it is appropriate to do this in another backport. I really was hoping that 3.3.x would be much more stable. I think we should limit ourselves to the lazy init in this PR.

Should we put a backport of https://github.com/cobbler/cobbler/pull/3677 in scope of this backport or not?

**Comment by tpw56j** [comment]
@SchoolGuy Ok, I’ll now additionally check that the cache really works as it should.

It seems to me that #3677 makes sense to include here.

**Comment by SchoolGuy** [comment]
@tpw56j This is a small PR. Give me half an hour to include it before you test.

**Comment by tpw56j** [comment]
@SchoolGuy I tested lazy start and everything seems to work as expected.

**Comment by SchoolGuy** [comment]
I am very relieved. Thanks a lot for that @tpw56j ! Now let's hope my colleagues find the time to give a formal review next week.

**Review by m-czernek — COMMENTED** [review]
Leaving a few comments. What I miss now is some mechanism to distinguish whether a search miss is the result of a collection not loaded, or a result of collection not existing.

This could be helped by e.g. by creating a new method that checks if memory loading has been finished or not, e.g.:

```
$ cobbler status --memory-loading-finished && echo $?
1 # loading is not finished
$ cobbler status --memory-loading-finished && echo $?
0 # loading is finished
```
Of course there are many possible solution, e.g. block the `cobbler ... find` method until the loading is finished, which might arguably be a better user experience, or modify the `cobbler ... find` output with some info about the loading process, etc. 

Not sure if you're planning to do this with a later PR.

**Comment by SchoolGuy** [comment]
@m-czernek As this is a backport only, I would kindly ask of it is okay for you if you document the issues you raised in an issue? The code and aproach in it is live on the unreleased development branch already for quite some time. If you have concerns with the backport itself I will gladly fix them.

**Comment by SchoolGuy** [comment]
Rebase to resolve conflict.

**Review by m-czernek — APPROVED** [review]
Ack, created a separate issue in https://github.com/cobbler/cobbler/issues/3722 - approving so that backporting can go forward.

**Pull request merged**